docs/rfcs/0002_templates_mvp.md

```diff
@@ -0,0 +1,339 @@
+
+-   Any organization or study admin may create or clone Templates/TemplateVersions
+
+## 🛠 Modifying Templates
+
+### UpdateTemplateMutation
+
+```graphql
+type UpdateTemplateInput {
+    name: String
+    description: String
+    owner: ID
+}
+```
+
+### UpdateTemplateVersionMutation
+
+```graphql
+type UpdateTemplateVersionInput {
+    description: String
+    fieldsSchema: JSONString
+    fieldsSchemaFile: Upload
+}
+```
+
+### Permissions
+
+-   Any organization or study administrator may update a Template or
+    TemplateVersion if their organization owns it
+
+## 🗑 Deleting Templates
+
+### DeleteTemplateMutation
+
+```graphql
+type DeleteTemplateInput {
+    id: ID!
+}
+```
+
+### DeleteTemplateVersionMutation
+
+```graphql
+type DeleteTemplateVersionInput {
+    id: ID!
+}
+```
+
+### Permissions
+
+### Permissions
+
+-   Any organization or study administrator may delete a Template or
+    TemplateVersion if their organization owns it
+
+## 📦 Downloading Templates
+
+### Download Study Templates
+
+Probably the easiest way to download templates since it bundles everything
+into an archive. You will get an archive containing:
+
+For each TemplateVersion in the study:
+
+-   TemplateVersion.submission_template_file: the TSV file that people will fill out and use for data submission to their study
+-   TemplateVersion.fields_schema: the template field definitions which serve as instructions on filling out the template
+
+    /download/study/<Study.kf_id>templates
+
+We should consider providing a multi-sheet Excel workbook with a sheet for each
+template's fields schema and a folder of TSVs with all of the submission
+templates.
+
+### Download Latest Template
+
+Download the latest version of template.
+
+```
+/download/template/<Template.id>
+```
+
+You would get an archive with two files:
+
+-   Template.latest_version.submission_template_file
+-   Template.latest_version.fields_schema
+
+### Download a Template Version
+
+Download a specific version of a template.
+
+```
+/download/template/<Template.id>/versions/<TemplateVersion.id>
+```
+
+You would get an archive with two files:
+
+-   TemplateVersion.submission_template_file
+-   TemplateVersion.fields_schema
+
+### Permissions
+
+-   Any user may download template files
+
+## Changes to Existing Functionality
+
+-   A `Study` will have a new field, `Study.template_versions`, a many to many
+    relationship to `TemplateVersion`.
+-   The `File.file_type` field will be replaced by a one to one foreign key
+    to a `TemplateVersion`. The `TemplateVersion` must be one of
+    templates in the Study's template versions.
+-   The study and file mutations should be updated accordingly as well
+
+## 💻 Implementation Considerations
+
+When thinking about implementing data submission templates, we should consider
+the entire development, test, and documentation lifecycle and some of the
+common painpoints encountered as they may inform a better implementation of our
+templates.
+
```

Comment:
We should think through how 'migrating' to a new template version happens.
We can create a new version of a template based on the latest version of a template, but how do we 'migrate' the files that use that template?
This could be easy if the change does something like remove a field. In this case, we can go through each document that used the old version and change it to the new one. This is possible because the document is valid as either the old or the new version.

However, it's a problem if we want to do something like add a field in a new version. All documents that use the old version will not be valid for the new version. This means we will have to add the new field to all the documents and upload new versions of them before we can change the document's template to the latest version.

So updating a document to match a new template version is possible, but it may not have the best user flow.
Something else to watch for is during this transition, some documents may use an old version while some may use a new version, but all documents will still be said to use the 'such and such template. This could be misleading unless we label documents with 'not latest version' badges or similar when browsing through documents and the templates they use.

Reply:
Yea, I think migration deserves a separate RFC or addendum. Haven't thought enough about it yet